### ISSUE TITLE
ARM64: Register restoration in asm_stack_check happens too late

### ISSUE THREAD

**Issue by corsix** [LuaJIT/LuaJIT#1068]
The "Restore temp register" part of ARM64's `asm_stack_check` happens on the wrong side of the `emit_cond_branch`, meaning that if the branch is taken, the temp register is not restored. Potential fix at https://github.com/corsix/LuaJIT/commits/arm64-stack-check.

This is incredibly hard to hit, even when trying to hit it deliberately, as in the setting of `asm_head_side`, it requires that `pallow` be the empty set, which in turn requires that `allow & RSET_GPR` be the empty set, which in turn makes it extremely easy to hit `LJ_TRERR_NYICOAL`. I managed to hit it with the following:
```lua
local function g(isum)
  for _ = 0.01, 0.6, 0.01 do end
  local A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V
  local A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V
  assert(0 <= isum and isum <= 50, isum) -- Fails, for example with: 43487425.5
end
local function f(i)
  for _ = 0.01, 0.6, 0.01 do
    local i1 = i*i; local i2 = i1*i1; local i3 = i2*i2; local i4 = i3*i3
    local i5 = i4*i4; local i6 = i5*i5; local i7 = i6*i6; local i8 = i7*i7
    local i9 = i8*i8; local i10 = i9*i9; local i11 = i10*i10; local i12 = i11*i11
    local i13 = i12*i12; local i14 = i13*i13; local i15 = i14*i14; local i16 = i15*i15
    local i17 = i16*i16; local i18 = i17*i17; local i19 = i18*i18; local i20 = i19*i19
    local i21 = i20*i20; local i22 = i21*i21; local i23 = i22*i22; local i24 = i23*i23
    local i25 = i24*i24; local i26 = i25*i25
    if i26 ~= 0 then
        -- Bunch of stuff to spill the `i` locals:
        local j = #{}
        local j1 = j+j; local j2 = j1+j1; local j3 = j2+j2; local j4 = j3+j3
        local j5 = j4+j4; local j6 = j5+j5; local j7 = j6+j6; local j8 = j7+j7
        local j9 = j8+j8; local j10 = j9+j9; local j11 = j10+j10; local j12 = j11+j11
        local j13 = j12+j12; local j14 = j13+j13; local j15 = j14+j14; local j16 = j15+j15
        local j17 = j16+j16; local j18 = j17+j17; local j19 = j18+j18; local j20 = j19+j19
        -- Back to normal:
        local isum = 0.5+i19+i1+i2+i3+i4+i5+i6+i7+i8+i9+i10+i11+i12+i13+i14+i15+i16+i17+i18+i20+i21+i22+i23+i24+i25+i26
        g(isum)
    end
  end
end
f(0)
f(1)
coroutine.wrap(f)(1)
```

**Comment by corsix** [comment]
I think that this code was inspired by the ARM32 code, so there's likely the same problem there.

However, the handling of REF_BASE in ARM32 is slightly funky; there's no good way to load from `GL->jit_base` on ARM32, so it prefers to maintain its own spill slot for REF_BASE. ARM64 doesn't have this problem; it maintains a permanent GL register, making loads from `GL->jit_base` just as cheap as loads from a spill slot. We can therefore bring the ARM64 backend in line with all non-ARM32 backends and allow `emit_canremat(REF_BASE)`, with corresponding changes to `asm_retf`, `asm_stack_check`, `asm_head_root_base`, and `asm_head_side_base`. While in the area, it looks like `asm_head_lreg` is dead code for ARM64, given that `emit_canremat(ASMREF_L)` is already true. Commit for this tidy-up at https://github.com/corsix/LuaJIT/commits/arm64-remat-base.

**Comment by MikePall** [comment]
I've opted for rematerialization from GL. Thanks!

I've fixed the ARM (32 bit) issue, too (v2.0 + v2.1).

There was a reason for 2bc63bb6affdb378c1698bd0f97bacb286a61a6f and it was not ARM-specific. I'll have to dig into my mail archives for that.

**Comment by MikePall** [comment]
OK, I found the original bug report and [a -jdump, which shows the problem](https://gist.github.com/MikePall/ae5efab360a6ca7e572b86b4d4ff7b40). No testcase, but the interim fix resolved the issue.
This should be moved to a separate issue, if anyone comes up with a testcase and a better fix.
